### PR TITLE
1823: Mailing list bot should include link to comment on the Git forge

### PR DESCRIPTION
**Pull request by zhaosongzs**
This patch will add the ability to attach comment/reviewComment/review link to the email generated by mlbridge bot

The implementation for GitHub is straightforward, we can extract the URL for a specific comment or review from the JSON response using the `html_url` field and include it in the email. 

For GitLab, we will need to generate the URL using the `id` field in `note,` since GitLab does not provide an field like `html_url` in its JSON responses.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1823](https://bugs.openjdk.org/browse/SKARA-1823): Mailing list bot should include link to comment on the Git forge

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara pull/1478/head:pull/1478` \
`$ git checkout pull/1478`

Update a local copy of the PR: \
`$ git checkout pull/1478` \
`$ git pull https://git.openjdk.org/skara pull/1478/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1478`

View PR using the GUI difftool: \
`$ git pr show -t 1478`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1478.diff">https://git.openjdk.org/skara/pull/1478.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — COMMENTED** [review]
When publishing URLs in emails, we rewrite them to be (somewhat) forge agnostic. If you look in any PR email, you will not that it says "git.openjdk.org" and not github.com. This is a service we host that redirects to "github.com/openjdk". If we were to ever move to a different provider, we can change that rewrite service. Because of this, we can't use the URLs in the PR object for emails. We need to generate them.

Since we don't need these URLs every time we fetch comments or reviews, I think it would make sense to let the PullRequest handle generation of these URLs, something like `PullRequest.reviewUrl(Review)` and `PullRequest.commentUrl(Comment)`, similar to the existing `PullRequest.changeUrl`.

**Comment by zhaosongzs** [comment]
> When publishing URLs in emails, we rewrite them to be (somewhat) forge agnostic. If you look in any PR email, you will not that it says "git.openjdk.org" and not github.com. This is a service we host that redirects to "github.com/openjdk". If we were to ever move to a different provider, we can change that rewrite service. Because of this, we can't use the URLs in the PR object for emails. We need to generate them.
> 
> Since we don't need these URLs every time we fetch comments or reviews, I think it would make sense to let the PullRequest handle generation of these URLs, something like `PullRequest.reviewUrl(Review)` and `PullRequest.commentUrl(Comment)`, similar to the existing `PullRequest.changeUrl`.

Oh, got it. I will fix the problems. Thx!

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java [hunk 1]
```diff
@@ -305,6 +305,20 @@ static String composeReplyFooter(PullRequest pr) {
         return "PR: " + pr.webUrl();
     }
 
+    static String composeReplyFooter(PullRequest pr, Comment comment, ReviewComment reviewComment, Review review) {
```
Comment (erikj79) [inline bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:308]: I think you should just split this into 3 separate methods. I can't see any caller that uses more than one of Comment, ReviewComment or Review.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java [hunk 2]
```diff
@@ -354,6 +354,9 @@ List<Email> generateNewEmails(List<Email> sentEmails, Duration cooldown, Reposit
                 var newFooterFragments = Stream.of(item.footer().split("\n\n"))
                                                .filter(line -> !includedFooterFragments.contains(line))
                                                .collect(Collectors.toList());
+                if(!footer.isEmpty() && !newFooterFragments.isEmpty()){
```
Comment (erikj79) [inline bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ReviewArchive.java:357]:
```suggestion
                if (!footer.isEmpty() && !newFooterFragments.isEmpty()) {
```

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java [hunk 3]
```diff
@@ -305,6 +305,20 @@ static String composeReplyFooter(PullRequest pr) {
...
+        var footer = new StringBuilder();
+        if (comment != null) {
+            footer.append("Comment: " + pr.commentUrl(comment).toString());
```
Comment (erikj79) [inline bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:311]:
```suggestion
            footer.append("PR Comment: " + pr.commentUrl(comment).toString());
```

**Comment by zhaosongzs** [comment]
This pr is still waiting for review.

**Comment by zhaosongzs** [comment]
Thanks for the review!
/integrate

**Pull request closed without merging**